pkg/querier/worker/scheduler_processor.go

```diff
@@ -184,18 +217,38 @@ func (sp *schedulerProcessor) runRequest(ctx context.Context, logger log.Logger,
 		}
 	}
 
+	result := &frontendv2pb.QueryResultRequest{
+		QueryID: queryID,
+		Response: &frontendv2pb.QueryResultRequest_HttpResponse{
+			HttpResponse: response,
+		},
+		Stats: stats,
+	}
+
+	sp.reply(ctx, logger, frontendAddress, result)
+}
+
+func (sp *schedulerProcessor) reply(ctx context.Context, logger log.Logger, frontendAddress string, result *frontendv2pb.QueryResultRequest) {
 	runPoolWithBackoff(
 		ctx,
 		logger,
 		sp.frontendPool,
 		frontendAddress,
 		func(c client.PoolClient) error {
 			// Response is empty and uninteresting.
-			_, err := c.(frontendv2pb.FrontendForQuerierClient).QueryResult(ctx, &frontendv2pb.QueryResultRequest{
-				QueryID:      queryID,
-				HttpResponse: response,
-				Stats:        stats,
-			})
+			_, err := c.(frontendv2pb.FrontendForQuerierClient).QueryResult(ctx, result)
+			if status, ok := status.FromError(err); ok {
+				if status.Code() == codes.ResourceExhausted {
+					level.Error(logger).Log("msg", "response larger than max message size", "maxMessageSize", sp.maxMessageSize)
+					result.Response = &frontendv2pb.QueryResultRequest_HttpResponse{
+						HttpResponse: &httpgrpc.HTTPResponse{
+							Code: http.StatusRequestEntityTooLarge,
+							Body: []byte(err.Error()),
+						},
+					}
+
+				}
+			}
```

Comment:
I've learned there's a [size](https://pkg.go.dev/github.com/gogo/protobuf/plugin/size?utm_source=godoc) plugin. So maybe we should use that.